### PR TITLE
Dataset query editor

### PR DESCRIPTION
**Pull request by kulyk**
Once a saved question is turned into a dataset, there is no way to edit the underlying query. It's intentional, so datasets behave as raw tables opened in simple mode (see #18993 and #19101).

This PR introduces a dataset editor page available on `/dataset/:id-:slug/query`, where people can edit their dataset queries

### To Verify

For GUI and native questions:

1. Turn a question into a dataset (following the steps from #18702)
2. Open the dataset details sidebar (click the dataset name while being on the `/dataset/:id-:slug` page
3. Click "Edit query definition"
4. Make sure you are navigated to a new page and it looks as expected
    * the global navbar is not shown
    * there is a query editor (notebook for structured queries and native editor for native queries)
    * there is a results table
    * you can resize the editor using a handle
5. Edit and re-run the query and click "Cancel" in the top right
    * you should be navigated back to `/dataset/:id-:slug`
    * the results table should have the results for the original query
    * open the editor again, you should  see the original query without any changes
6. Edit the query and click "Save changes"
    * you should be navigated back to `/dataset/:id-:slug`
    * the results table should now have the results for the updated query
    * open the editor again, you should  see the updated query

### Demo

**Structured Query**

https://user-images.githubusercontent.com/17258145/143591558-91f79ed9-59d1-4c23-a6ed-f6a57962c072.mov

**Native Query**

https://user-images.githubusercontent.com/17258145/143591571-895b2c95-f9de-4e00-9bbd-fdbb829c08cc.mov

### CODE REVIEW

**Comment by mazameli** [comment]
It's pretty cool actually that you can just click and drag the columns in the results preview to reorder them, and I don't want to lose that. But we _shouldn't_ have any of the other drill-through actions etc. as you see in this video, like sorting or filtering.

https://www.loom.com/share/2c6bcb86353f417f89985482c31ef31f

**Comment by mazameli** [comment]
When resizing the preview window, I'd expect the editor UI above it to stay in place, not scroll up as it's doing in this recording: https://www.loom.com/share/1d3931aa002f489db7b39cce87cd33a5

**Comment by mazameli** [comment]
This might just be a nested query bug, but I'm going to mention it here because it seems like a thing that'll come up often: if my dataset query definition contains a filter on a date column, like this one...

image

... then trying to summarize on that dataset by grouping on that datetime column results in an error:

<details>
<summary>Error text</summary>
Column "source.CREATED_AT" not found; SQL statement: -- Metabase:: userID: 1 queryType: MBQL queryHash: 6c520dcf3ca10f4b7320f44a4469ecf6c5f8af8de8a9dbfa5e3561eff0c1582c SELECT parsedatetime(formatdatetime("source"."CREATED_AT", 'yyyyMM'), 'yyyyMM') AS "CREATED_AT", count(*) AS "count" FROM (SELECT "PUBLIC"."PEOPLE"."ID" AS "ID", "PUBLIC"."PEOPLE"."ADDRESS" AS "ADDRESS", "PUBLIC"."PEOPLE"."EMAIL" AS "EMAIL", "PUBLIC"."PEOPLE"."NAME" AS "NAME", "PUBLIC"."PEOPLE"."CITY" AS "CITY", "PUBLIC"."PEOPLE"."SOURCE" AS "SOURCE" FROM "PUBLIC"."PEOPLE" WHERE ("PUBLIC"."PEOPLE"."CREATED_AT" >= parsedatetime(year(dateadd('year', CAST(-3 AS long), now())), 'yyyy') AND "PUBLIC"."PEOPLE"."CREATED_AT" < parsedatetime(year(now()), 'yyyy') AND "PUBLIC"."PEOPLE"."STATE" = ?)) "source" GROUP BY parsedatetime(formatdatetime("source"."CREATED_AT", 'yyyyMM'), 'yyyyMM') ORDER BY parsedatetime(formatdatetime("source"."CREATED_AT", 'yyyyMM'), 'yyyyMM') ASC [42122-197]
</details>

image

**Comment by mazameli** [comment]
The column heading menu for SQL-based datasets looks wrong. 

image

This is likely because the menu is different for SQL questions than GUI ones normally, as you can see on this saved SQL question:

image

**Comment by mazameli** [comment]
I can't seem to open the variables side panel in the SQL-based dataset editor. I _think_ I should be able to, since there are some variables that would make sense with a SQL dataset, like using the CTE syntax, e.g. {{#123}} to insert/reference the results of another query within the current one.

image

**Comment by mazameli** [comment]
Actually, it looks like the data reference and Snippets menus aren't working either, and I would definitely expect them to here: 

image

**Comment by kulyk** [comment]
Huge thanks for your review @mazameli 🙏  Mostly every comment is addressed, could please have one more look?

- [x] [disable drill-through actions in the editor](https://github.com/metabase/metabase/pull/19103#issuecomment-981931919)
- [x] [disable scroll behavior when resizing the editor](https://github.com/metabase/metabase/pull/19103#issuecomment-981935878)
- [x] [fix couldn't open native sidebars (data reference, variables, etc.)](https://github.com/metabase/metabase/pull/19103#issuecomment-981969529)
- [ ] [summarizing by filtered datetime column results in a broken query](https://github.com/metabase/metabase/pull/19103#issuecomment-981948140) — seems to be a nested query bug as you mentioned, would make sense to tackle it separately
- [ ] [column headings for SQL-based datasets look wrong](https://github.com/metabase/metabase/pull/19103#issuecomment-981965958) — was broken earlier, so I'd like to open a separate PR for that

**Comment by mazameli** [comment]
A nice touch would be to give these column headings `cursor: grab;` to show you can reorder them.

image

**Review by mazameli — APPROVED** [review]
Looks like the things I spotted are all fixed. 👍🏼 As I commented, a nice touch would be to add `cursor: grab` to the column headings of the results area of the editor.

**Review by daltojohnso — APPROVED** [review]
code looks good. one missed translation. i also noticed a bug when trying to update a dataset based on a native question. doesn't seem super related to this PR, but here's a recording:

https://user-images.githubusercontent.com/13057258/144289660-743cda6d-1c45-467c-8977-7f363c70ee89.mov

File: frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    await props.onSave(dataset.card());
+    setQueryBuilderMode("view");
+  };
+
+  const sidebar = getSidebar(props);
+
+  return (
+    <React.Fragment>
+      <EditBar
+        title={`You're editing ${dataset.displayName()}`}
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx]: Thanks for always being on a lookout for missing translations @daltojohnso 
You literally saved me from dozens of them already. Kudos 🙏

[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx]: Also fixed the issue with template tags for native datasets. Thanks again!

**Pull request merged**